File: labs/AutomataTheory/lab2/code/my_bot_v2.pas
unit my_bot_v2;

interface

uses Math;

const
  ROCK = 1;
  PAPER = 2;
  SCISSORS = 3;

  WIN = 1;
  LOSE = -1;
  DRAW = 0;

  MAX_HISTORY = 1000;
  MAX_PATTERN_LENGTH = 10;
  ANALYSIS_WINDOW = 15;
  MIN_STRATEGY_DATA = 3;
  RECENT_WINDOW = 20;

type
  THistory = record
    opponentMoves: array[0..MAX_HISTORY-1] of integer;
    ourMoves: array[0..MAX_HISTORY-1] of integer;
    outcomes: array[0..MAX_HISTORY-1] of integer;
    length: integer;
  end;

  TStrategyStats = record
    correct: integer;
    total: integer;
  end;

var
  history: THistory;
  patternStats, frequencyStats: TStrategyStats;

function beats(move: integer): integer;
function getCounterMove(predictedMove: integer): integer;
function determineOutcome(ourMove, opponentMove: integer): integer;
function analyzePatterns(patternLength: integer): integer;
function analyzeFrequency(): integer;
function choose(previousOpponentChoice: integer): integer;
procedure onGameStart();
procedure setParameters(setCount, winsPerSet: integer);
procedure onGameEnd();

implementation

var
  current_round: integer;

function beats(move: integer): integer;
begin
  case move of
    ROCK: beats := PAPER;
    PAPER: beats := SCISSORS;
    SCISSORS: beats := ROCK;
    else beats := PAPER;
  end;
end;

function getCounterMove(predictedMove: integer): integer;
begin
  getCounterMove := beats(predictedMove);
end;

function determineOutcome(ourMove, opponentMove: integer): integer;
begin
  if ourMove = opponentMove then
    determineOutcome := DRAW
  else if ((ourMove = ROCK) and (opponentMove = SCISSORS)) or
          ((ourMove = PAPER) and (opponentMove = ROCK)) or
          ((ourMove = SCISSORS) and (opponentMove = PAPER)) then
    determineOutcome := WIN
  else
    determineOutcome := LOSE;
end;

procedure calculateRecentWinrates(var patternWinrate, freqWinrate: real);
var
  patternRecent, patternRecentTotal: integer;
  freqRecent, freqRecentTotal: integer;
begin
  patternRecent := 0;
  patternRecentTotal := 0;
  freqRecent := 0;
  freqRecentTotal := 0;

  if patternStats.total > 0 then
  begin
    patternRecentTotal := Min(patternStats.total, RECENT_WINDOW);
    patternRecent := patternStats.correct;
    if patternStats.total > RECENT_WINDOW then
      patternRecent := Round(patternStats.correct * RECENT_WINDOW / patternStats.total);
  end;

  if frequencyStats.total > 0 then
  begin
    freqRecentTotal := Min(frequencyStats.total, RECENT_WINDOW);
    freqRecent := frequencyStats.correct;
    if frequencyStats.total > RECENT_WINDOW then
      freqRecent := Round(frequencyStats.correct * RECENT_WINDOW / frequencyStats.total);
  end;

  if patternRecentTotal >= MIN_STRATEGY_DATA then
    patternWinrate := patternRecent / patternRecentTotal
  else
    patternWinrate := -1.0;

  if freqRecentTotal >= MIN_STRATEGY_DATA then
    freqWinrate := freqRecent / freqRecentTotal
  else
    freqWinrate := -1.0;
end;

function selectBestPrediction(patternPred, freqPred: integer;
                              patternWinrate, freqWinrate: real): integer;
begin
  if (patternPred <> -1) and (freqPred <> -1) then
  begin
    if patternWinrate > freqWinrate then
      selectBestPrediction := patternPred
    else
      selectBestPrediction := freqPred;
  end
  else if patternPred <> -1 then
    selectBestPrediction := patternPred
  else if freqPred <> -1 then
    selectBestPrediction := freqPred
  else
    selectBestPrediction := -1;
end;

function getFallbackPrediction(patternPred, freqPred: integer): integer;
var
  recentMoves: array[0..2] of integer;
  count, i, start: integer;
begin
  if patternPred <> -1 then
  begin
    getFallbackPrediction := patternPred;
    exit;
  end;
  if freqPred <> -1 then
  begin
    getFallbackPrediction := freqPred;
    exit;
  end;

  if history.length >= 2 then
  begin
    count := 0;
    start := Max(0, history.length - 3);
    for i := start to history.length - 1 do
    begin
      if (history.opponentMoves[i] >= ROCK) and (history.opponentMoves[i] <= SCISSORS) then
      begin
        recentMoves[count] := history.opponentMoves[i];
        Inc(count);
        if count >= 3 then break;
      end;
    end;
    if count > 0 then
    begin
      getFallbackPrediction := getCounterMove(recentMoves[count - 1]);
      exit;
    end;
  end;

  getFallbackPrediction := (history.length mod 3) + 1;
end;

function analyzePatterns(patternLength: integer): integer;
var
  pLen, patternStart, i, j: integer;
  match: boolean;
begin
  if history.length < patternLength + 1 then
  begin
    analyzePatterns := -1;
    exit;
  end;

  for pLen := patternLength downto 2 do
  begin
    if history.length < pLen + 1 then continue;
    patternStart := history.length - pLen;

    for i := history.length - pLen - 1 downto pLen do
    begin
      match := true;
      for j := 0 to pLen - 1 do
      begin
        if history.opponentMoves[i - pLen + j] <> history.opponentMoves[patternStart + j] then
        begin
          match := false;
          break;
        end;
      end;
      if match then
      begin
        analyzePatterns := history.opponentMoves[i];
        exit;
      end;
    end;
  end;

  analyzePatterns := -1;
end;

function analyzeFrequency(): integer;
var
  rockCount, paperCount, scissorsCount: real;
  window, startIdx, i: integer;
begin
  if history.length < 3 then
  begin
    analyzeFrequency := -1;
    exit;
  end;

  rockCount := 0.0;
  paperCount := 0.0;
  scissorsCount := 0.0;

  window := Min(history.length, 20);
  startIdx := history.length - window;

  for i := startIdx to history.length - 1 do
  begin
    case history.opponentMoves[i] of
      ROCK: rockCount := rockCount + (0.5 + (i - startIdx) / window);
      PAPER: paperCount := paperCount + (0.5 + (i - startIdx) / window);
      SCISSORS: scissorsCount := scissorsCount + (0.5 + (i - startIdx) / window);
    end;
  end;

  if (rockCount >= paperCount) and (rockCount >= scissorsCount) then
    analyzeFrequency := ROCK
  else if paperCount >= scissorsCount then
    analyzeFrequency := PAPER
  else
    analyzeFrequency := SCISSORS;
end;

function choose(previousOpponentChoice: integer): integer;
var
  ourMove, prediction, outcome: integer;
  lastMove, sameCount, windowStart, i: integer;
  counts: array[0..2] of integer;
  maxIdx: integer;
  drawCount: integer;
  patternPred, freqPred: integer;
  patternWinrate, freqWinrate: real;
begin
  if previousOpponentChoice = 0 then
  begin
    ourMove := PAPER;
    history.ourMoves[0] := PAPER;
    history.opponentMoves[0] := 0;
    history.length := 1;
    current_round := 1;
    choose := PAPER;
    exit;
  end;

  history.opponentMoves[history.length - 1] := previousOpponentChoice;

  outcome := determineOutcome(history.ourMoves[history.length - 1], previousOpponentChoice);
  history.outcomes[history.length - 1] := outcome;

  // Update stats
  patternPred := analyzePatterns(MAX_PATTERN_LENGTH);
  if patternPred <> -1 then
  begin
    Inc(patternStats.total);
    if patternPred = previousOpponentChoice then Inc(patternStats.correct);
  end;

  freqPred := analyzeFrequency();
  if freqPred <> -1 then
  begin
    Inc(frequencyStats.total);
    if freqPred = previousOpponentChoice then Inc(frequencyStats.correct);
  end;

  // Choose strategy
  patternPred := analyzePatterns(MAX_PATTERN_LENGTH);
  freqPred := analyzeFrequency();
  calculateRecentWinrates(patternWinrate, freqWinrate);

  prediction := selectBestPrediction(patternPred, freqPred, patternWinrate, freqWinrate);

  if prediction = -1 then
    prediction := getFallbackPrediction(patternPred, freqPred);

  ourMove := getCounterMove(prediction);

  // Anti-repetition
  if history.length >= 5 then
  begin
    lastMove := history.ourMoves[history.length - 1];
    sameCount := 1;
    for i := history.length - 2 downto Max(0, history.length - 10) do
    begin
      if history.ourMoves[i] = lastMove then
        Inc(sameCount)
      else
        break;
    end;

    if (sameCount >= 3) and (ourMove = lastMove) then
    begin
      FillChar(counts, SizeOf(counts), 0);
      windowStart := Max(0, history.length - 5);
      for i := windowStart to history.length - 1 do
      begin
        if (history.opponentMoves[i] >= ROCK) and (history.opponentMoves[i] <= SCISSORS) then
          Inc(counts[history.opponentMoves[i] - 1]);
      end;

      maxIdx := 0;
      if counts[1] > counts[maxIdx] then maxIdx := 1;
      if counts[2] > counts[maxIdx] then maxIdx := 2;

      prediction := maxIdx + 1;
      ourMove := getCounterMove(prediction);

      if ourMove = lastMove then
        ourMove := (ourMove mod 3) + 1;
    end;
  end;

  // Anti-draw cycle
  if history.length >= 3 then
  begin
    drawCount := 0;
    for i := history.length - 3 to history.length - 1 do
    begin
      if history.outcomes[i] = DRAW then Inc(drawCount);
    end;

    if drawCount >= 3 then
    begin
      FillChar(counts, SizeOf(counts), 0);
      for i := Max(0, history.length - 8) to history.length - 1 do
      begin
        if (history.opponentMoves[i] >= ROCK) and (history.opponentMoves[i] <= SCISSORS) then
          Inc(counts[history.opponentMoves[i] - 1]);
      end;

      maxIdx := 0;
      if counts[1] > counts[maxIdx] then maxIdx := 1;
      if counts[2] > counts[maxIdx] then maxIdx := 2;

      prediction := maxIdx + 1;
      ourMove := getCounterMove(prediction);
    end;
  end;

  if history.length < MAX_HISTORY then
  begin
    history.ourMoves[history.length] := ourMove;
    history.opponentMoves[history.length] := 0;
    Inc(history.length);
  end;

  Inc(current_round);
  choose := ourMove;
end;

procedure onGameStart();
begin
  history.length := 0;
  patternStats.correct := 0;
  patternStats.total := 0;
  frequencyStats.correct := 0;
  frequencyStats.total := 0;
  current_round := 0;
end;

procedure setParameters(setCount, winsPerSet: integer);
begin
end;

procedure onGameEnd();
begin
end;

end.

File: labs/AutomataTheory/lab2/code/my_bot_v2.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define ROCK 1
#define PAPER 2
#define SCISSORS 3

#define WIN 1
#define LOSE -1
#define DRAW 0

#define MAX_HISTORY 1000
#define MAX_PATTERN_LENGTH 10
#define LOSE_STREAK_THRESHOLD 5  // СМЕНА СТРАТЕГИИ ТОЛЬКО ПРИ 5 ПОРАЖЕНИЯХ ПОДРЯД

typedef struct {
    int opponentMoves[MAX_HISTORY];
    int ourMoves[MAX_HISTORY];
    int outcomes[MAX_HISTORY];
    int length;
} THistory;

THistory history;
int current_round = 0;
int current_lose_streak = 0;
int current_strategy = 1;  // 1 = стандартная, 2 = модифицированная

int beats(int move) {
    switch (move) {
        case ROCK:     return PAPER;
        case PAPER:    return SCISSORS;
        case SCISSORS: return ROCK;
        default:       return PAPER;
    }
}

int getCounterMove(int predictedMove) { return beats(predictedMove); }

int determineOutcome(int ourMove, int opponentMove) {
    if (ourMove == opponentMove) return DRAW;
    if ((ourMove == ROCK && opponentMove == SCISSORS) ||
        (ourMove == PAPER && opponentMove == ROCK) ||
        (ourMove == SCISSORS && opponentMove == PAPER))
        return WIN;
    return LOSE;
}

/* ============= УСИЛЕННЫЙ ПРОГНОЗ: ТОЧНОСТЬ + АДАПТАЦИЯ ============= */

/**
 * Анализ 1: ВЗАИМОДЕЙСТВИЕ - ищем паттерны "наш ход + ход противника"
 * УСКОРЕНИЕ: ищем только в последних 5 парах
 */
int analyzeInteractionPatterns(int maxLen) {
    if (history.length < maxLen + 1) return -1;
    int search_start = (history.length > 5) ? history.length - 5 : 0;
    for (int len = maxLen; len >= 2; len--) {
        if (history.length < len + 1) continue;
        int cur_start = history.length - len;
        for (int i = search_start; i < history.length - len; i++) {
            int match = 1;
            for (int j = 0; j < len; j++) {
                if (history.ourMoves[i + j] != history.ourMoves[cur_start + j] ||
                    history.opponentMoves[i + j] != history.opponentMoves[cur_start + j]) {
                    match = 0; break;
                }
            }
            if (match) return history.opponentMoves[i + len];
        }
    }
    return -1;
}

/**
 * Анализ 2: РЕАКЦИЯ НА ИСХОД
 */
int analyzeOutcomeReaction() {
    if (history.length < 3) return -1;
    int last_outcome = history.outcomes[history.length - 2];
    int last_opp = history.opponentMoves[history.length - 2];
    if (last_outcome == -WIN) return beats(last_opp);  // Он проиграл → сменит
    if (last_outcome == WIN)  return last_opp;         // Он выиграл → повторит
    return -1;
}

/**
 * Анализ 3: ПАТТЕРНЫ ПРОТИВНИКА
 */
int analyzePatterns(int maxLen) {
    if (history.length < maxLen + 1) return -1;
    for (int len = maxLen; len >= 2; len--) {
        if (history.length < len + 1) continue;
        int cur_start = history.length - len;
        for (int i = len; i < history.length - len; i++) {
            int match = 1;
            for (int j = 0; j < len; j++) {
                if (history.opponentMoves[i + j] != history.opponentMoves[cur_start + j]) {
                    match = 0; break;
                }
            }
            if (match) return history.opponentMoves[i + len];
        }
    }
    return -1;
}

/**
 * Анализ 3.5: ПОСЛЕДНИЙ ХОД ПРОТИВНИКА (если ничего нет)
 */
int predictLastMove() {
    if (history.length < 2) return -1;
    return history.opponentMoves[history.length - 1];
}

/**
 * Анализ 4: ЧАСТОТА — УСИЛЕННАЯ ВЕРСИЯ
 * Последние 3 хода — вес ×2
 */
int analyzeFrequency() {
    if (history.length < 2) return -1;
    double r = 0, p = 0, s = 0;
    int window = (history.length < 20) ? history.length : 20;
    int start = history.length - window;

    for (int i = start; i < history.length; i++) {
        double w = 0.5 + (i - start) / (double)window;
        if (i >= history.length - 3) w *= 2.0;  // УСИЛЕНИЕ ПОСЛЕДНИХ 3
        switch (history.opponentMoves[i]) {
            case ROCK:     r += w; break;
            case PAPER:    p += w; break;
            case SCISSORS: s += w; break;
        }
    }
    if (r >= p && r >= s) return ROCK;
    if (p >= s) return PAPER;
    return SCISSORS;
}

/**
 * Анализ 5: FALLBACK
 */
int getFallbackPrediction() {
    return (history.length % 3) + 1;
}

/**
 * ГЛАВНАЯ ФУНКЦИЯ ПРОГНОЗА — УСИЛЕННАЯ ИЕРАРХИЯ
 */
int getPrediction() {
    int pred;
    if ((pred = analyzeInteractionPatterns(MAX_PATTERN_LENGTH)) != -1) return pred;
    if ((pred = analyzeOutcomeReaction()) != -1) return pred;
    if ((pred = analyzePatterns(MAX_PATTERN_LENGTH)) != -1) return pred;
    if ((pred = predictLastMove()) != -1) return pred;  // НОВЫЙ УРОВЕНЬ
    if ((pred = analyzeFrequency()) != -1) return pred;
    return getFallbackPrediction();
}

/* ============= АДАПТАЦИЯ И СТРАТЕГИИ ============= */

void switchStrategyIfNeeded(int outcome) {
    if (outcome == LOSE) {
        current_lose_streak++;
        if (current_lose_streak >= LOSE_STREAK_THRESHOLD) {
            current_strategy = (current_strategy == 1) ? 2 : 1;
            current_lose_streak = 0;
        }
    } else {
        current_lose_streak = 0;
    }
}

int cycleMove(int move) {
    if (move == ROCK) return PAPER;
    if (move == PAPER) return SCISSORS;
    if (move == SCISSORS) return ROCK;
    return PAPER;
}

/* ============= ГЛАВНАЯ ЛОГИКА ============= */

int choose(int previousOpponentChoice) {
    current_round++;

    if (previousOpponentChoice == 0) {
        history.ourMoves[0] = PAPER;
        history.opponentMoves[0] = 0;
        history.length = 1;
        current_strategy = 1;
        current_lose_streak = 0;
        return PAPER;
    }

    history.opponentMoves[history.length - 1] = previousOpponentChoice;
    int outcome = determineOutcome(history.ourMoves[history.length - 1], previousOpponentChoice);
    history.outcomes[history.length - 1] = outcome;

    int prediction = getPrediction();
    int ourMove;

    if (current_strategy == 1) {
        ourMove = getCounterMove(prediction);

        // АНТИ-СТАГНАЦИЯ: 2 поражения подряд → цикл
        if (history.length >= 2 &&
            history.outcomes[history.length-1] == LOSE &&
            history.outcomes[history.length-2] == LOSE) {
            ourMove = cycleMove(ourMove);
        }

        // АНТИ-ПОВТОР: 3 одинаковых хода + 1 поражение → цикл
        if (history.length >= 3 &&
            history.ourMoves[history.length-1] == history.ourMoves[history.length-2] &&
            history.ourMoves[history.length-2] == history.ourMoves[history.length-3] &&
            history.outcomes[history.length-1] == LOSE) {
            ourMove = cycleMove(ourMove);
        }

    } else {
        static int sequence[10] = {SCISSORS, SCISSORS, PAPER, ROCK, SCISSORS,
                                   ROCK, PAPER, SCISSORS, PAPER, PAPER};
        static int idx = 0;
        if (idx < 6) {
            ourMove = getCounterMove(prediction);
        } else if (idx < 8) {
            // ПРОВОКАЦИЯ: контр к своему предыдущему
            ourMove = beats(history.ourMoves[history.length-1]);
        } else {
            ourMove = sequence[idx];
        }
        idx = (idx + 1) % 10;
    }

    switchStrategyIfNeeded(outcome);

    if (history.length < MAX_HISTORY) {
        history.ourMoves[history.length] = ourMove;
        history.opponentMoves[history.length] = 0;
        history.length++;
    }

    return ourMove;
}

void onGameStart() {
    memset(&history, 0, sizeof(history));
    current_round = 0;
    current_strategy = 1;
    current_lose_streak = 0;
}

void setParameters(int setCount, int winsPerSet) {
    // Не используется
}